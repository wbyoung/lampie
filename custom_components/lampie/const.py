"""Constants for the Lampie integration."""

from typing import Final

from homeassistant.const import ATTR_ENTITY_ID as _ATTR_ENTITY_ID

DOMAIN: Final = "lampie"

ATTR_BRIGHTNESS: Final = "brightness"
ATTR_COLOR: Final = "color"
ATTR_DURATION: Final = "duration"
ATTR_EFFECT: Final = "effect"
ATTR_ENTITY_ID: Final = _ATTR_ENTITY_ID
ATTR_EXPIRES_AT: Final = "expires_at"
ATTR_LED_CONFIG: Final = "leds"
ATTR_INDIVIDUAL: Final = "individual"
ATTR_NAME: Final = "name"
ATTR_NOTIFICATION: Final = "notification"
ATTR_STARTED_AT: Final = "started_at"
ATTR_TYPE: Final = "type"
ATTR_VALUE: Final = "value"

CONF_BRIGHTNESS: Final = "brightness"
CONF_COLOR: Final = "color"
CONF_DURATION: Final = "duration"
CONF_EFFECT: Final = "effect"
CONF_END_ACTION: Final = "end_action"
CONF_LED_CONFIG: Final = "led_config"
CONF_NAME: Final = "name"
CONF_PRIORITY: Final = "priority"
CONF_START_ACTION: Final = "start_action"
CONF_SWITCH_ENTITIES: Final = "switches"

INOVELLI_MODELS = {
    "LZW30-SN",  # red on/off switch
    "LZW31-SN",  # red dimmer
    "LZW36",  # red fan/light combo
    "VTM31-SN",  # white dimmer (2-in-1 switch/dimmer)
    "VTM35-SN",  # white fan
    "VZM30-SN",  # blue switch
    "VZM31-SN",  # blue 2-in-1 switch/dimmer
    "VZM35-SN",  # blue fan switch
    "VZM36",  # blue canopy module
    "VZW31-SN",  # red 2-in-1 dimmer
}


ZWAVE_EFFECT_PARAMETERS = {
    "LZW31-SN": 16,  # red dimmer
    "LZW30-SN": 8,  # red on/off switch
    "LZW36_light": 24,  # red fan/light combo
    "LZW36_fan": 25,  # red fan/light combo
    "VZW31-SN": 99,  # red 2-in-1 dimmer
    "VZW31-SN_individual_1": 64,
    "VZW31-SN_individual_2": 69,
    "VZW31-SN_individual_3": 74,
    "VZW31-SN_individual_4": 79,
    "VZW31-SN_individual_5": 84,
    "VZW31-SN_individual_6": 89,
    "VZW31-SN_individual_7": 94,
}

ZWAVE_EFFECT_MAPPING = {
    "LZW30-SN": {  # red on/off switch
        "CLEAR": 0,
        "AURORA": 4,
        "MEDIUM_BLINK": 3,
        "CHASE": 2,
        "FAST_CHASE": 2,
        "SLOW_CHASE": 3,
        "FAST_FALLING": 2,
        "MEDIUM_FALLING": 3,
        "SLOW_FALLING": 3,
        "OPEN_CLOSE": 4,
        "FAST_RISING": 2,
        "MEDIUM_RISING": 3,
        "SLOW_RISING": 3,
        "FAST_SIREN": 4,
        "SLOW_SIREN": 4,
        "SMALL_TO_BIG": 4,
    },
    "LZW31-SN": {  # red dimmer
        "AURORA": 4,
        "FAST_BLINK": 3,
        "MEDIUM_BLINK": 4,
        "SLOW_BLINK": 4,
        "CHASE": 2,
        "FAST_CHASE": 2,
        "SLOW_CHASE": 2,
        "FAST_FALLING": 2,
        "MEDIUM_FALLING": 2,
        "SLOW_FALLING": 2,
        "OPEN_CLOSE": 2,
        "PULSE": 5,
        "FAST_RISING": 2,
        "MEDIUM_RISING": 2,
        "SLOW_RISING": 2,
        "SLOW_SIREN": 2,
        "FAST_SIREN": 2,
        "SMALL_TO_BIG": 2,
    },
    "LZW36": {  # red fan/light combo
        "AURORA": 4,
        "FAST_BLINK": 3,
        "MEDIUM_BLINK": 4,
        "SLOW_BLINK": 4,
        "CHASE": 2,
        "FAST_CHASE": 2,
        "SLOW_CHASE": 2,
        "FAST_FALLING": 2,
        "MEDIUM_FALLING": 2,
        "SLOW_FALLING": 2,
        "OPEN_CLOSE": 2,
        "PULSE": 5,
        "FAST_RISING": 2,
        "MEDIUM_RISING": 2,
        "SLOW_RISING": 2,
        "SLOW_SIREN": 2,
        "FAST_SIREN": 2,
        "SMALL_TO_BIG": 2,
    },
}


TRACE: Final = 5
