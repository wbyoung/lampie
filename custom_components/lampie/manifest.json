{
  "domain": "lampie",
  "name": "Lampie",
  "codeowners": [
    "@wbyoung"
  ],
  "config_flow": true,
  "dependencies": ["mqtt"],
  "documentation": "https://github.com/wbyoung/lampie",
  "homekit": {},
  "iot_class": "calculated",
  "issue_tracker": "https://github.com/wbyoung/lampie/issues",
  "requirements": [],
  "ssdp": [],
  "version": "0.1.0",
  "zeroconf": []
}
