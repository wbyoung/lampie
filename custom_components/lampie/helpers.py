"""Lampie helpers."""

from __future__ import annotations

from collections.abc import AsyncGenerator
from contextlib import asynccontextmanager
import logging

from homeassistant.config_entries import ConfigEntry, ConfigEntryState
from homeassistant.core import HomeAssistant
from homeassistant.helpers import device_registry as dr
from homeassistant.helpers.device import async_device_info_to_link_from_entity
from homeassistant.util import slugify

from .const import CONF_PRIORITY, DOMAIN
from .types import LampieConfigEntry, LampieConfigEntryRuntimeData

_LOGGER = logging.getLogger(__name__)


def is_primary_for_switch(entry: LampieConfigEntry, switch_id: str) -> bool:
    priorities = entry.data.get(CONF_PRIORITY, {}).get(switch_id)

    if not priorities:
        return True
    return bool(priorities[0] == slugify(entry.title))


def lookup_device_info(
    hass: HomeAssistant, entry: LampieConfigEntry, switch_id: str
) -> dr.DeviceInfo:
    device_info = async_device_info_to_link_from_entity(hass, switch_id)

    if not device_info:
        _LOGGER.warning(
            "skipping creation of sensors for %s on %s because an associated "
            "device could not be found",
            switch_id,
            entry.title,
        )

    return device_info


def get_other_entries(
    config_entry: ConfigEntry, *, hass: HomeAssistant
) -> list[ConfigEntry]:
    return [
        entry
        for entry in hass.config_entries.async_entries(DOMAIN)
        if entry is not config_entry
    ]


@asynccontextmanager
async def auto_reload_disabled(  # noqa: RUF029
    other_entries: list[ConfigEntry],
) -> AsyncGenerator[None]:
    targets: list[LampieConfigEntryRuntimeData] = [
        entry.runtime_data for entry in other_entries if hasattr(entry, "runtime_data")
    ]

    for runtime_data in targets:
        runtime_data.auto_reload_enabled = False

    try:
        yield
    finally:
        for runtime_data in targets:
            runtime_data.auto_reload_enabled = True


@asynccontextmanager
async def unloaded(
    other_entries: list[ConfigEntry],
    *,
    hass: HomeAssistant,
) -> AsyncGenerator[None]:
    other_entries = [  # filter to only the loaded entries
        entry for entry in other_entries if entry.state is ConfigEntryState.LOADED
    ]
    for entry in other_entries:
        await hass.config_entries.async_unload(entry.entry_id)

    yield

    for entry in other_entries:
        assert entry.state is ConfigEntryState.NOT_LOADED
        await hass.config_entries.async_setup(entry.entry_id)
