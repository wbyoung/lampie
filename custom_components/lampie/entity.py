"""Base support for Lampie entities."""

from __future__ import annotations

from abc import ABC, abstractmethod
from collections.abc import Callable
import logging
from typing import Any

from homeassistant.helpers.device_registry import DeviceEntryType, DeviceInfo
from homeassistant.helpers.entity import EntityDescription
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import DOMAIN
from .coordinator import LampieUpdateCoordinator
from .orchestrator import LampieOrchestrator

_LOGGER = logging.getLogger(__name__)


class LampieEntity[DataT](ABC, CoordinatorEntity[LampieUpdateCoordinator]):
    """Lampie entity class."""

    _attr_has_entity_name = True
    entity_description: LampieEntityDescription

    def __init__(
        self,
        *,
        description: LampieEntityDescription,
        coordinator: LampieUpdateCoordinator,
    ) -> None:
        """Initialize the sensor."""
        super().__init__(coordinator)

        self.entity_description = description

        self._attr_unique_id = f"{coordinator.config_entry.entry_id}_{description.key}"
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, coordinator.config_entry.entry_id)},
            name=coordinator.config_entry.title,
            entry_type=DeviceEntryType.SERVICE,
        )

    @property
    @abstractmethod
    def data(self) -> DataT:
        """Get data from coordinator."""

    @property
    def orchestrator(self) -> LampieOrchestrator:
        """Get the Lampie orchestrator."""
        config_entry = self.coordinator.config_entry
        orchestrator: LampieOrchestrator = config_entry.runtime_data.orchestrator
        return orchestrator

    @property
    def available(self) -> bool:
        """Return if entity is available."""
        return super().available and self.data is not None

    @property
    def extra_state_attributes(self) -> dict[str, Any] | None:
        """Return the state attributes."""
        return self.entity_description.attrs_fn(self.data)


class LampieDistributedEntity[DataT](LampieEntity[DataT]):
    """Lampie entity class for those distributed across several switches."""

    _attr_has_entity_name = True
    entity_description: LampieEntityDescription

    def __init__(
        self,
        *,
        description: LampieEntityDescription,
        coordinator: LampieUpdateCoordinator,
        switch_id: str,
        switch_device_info: DeviceInfo,
    ) -> None:
        """Initialize the sensor."""
        super().__init__(description=description, coordinator=coordinator)

        self.switch_id = switch_id

        id_tuple = next(iter(switch_device_info["identifiers"]))
        id_slug = "_".join(id_tuple)

        self._attr_unique_id = f"{id_slug}_{description.key}"
        self._attr_device_info = switch_device_info


class LampieEntityDescription[DataT](EntityDescription):
    """Class describing Lampie entities."""

    value_fn: Callable[
        [DataT],
        str | int | float | None,
    ]
    attrs_fn: Callable[
        [DataT],
        dict[str, Any] | None,
    ] = lambda _: None
