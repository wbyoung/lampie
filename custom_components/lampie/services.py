"""Support for the Lapmpie services."""

from functools import partial
from typing import Any, Final

from homeassistant.const import ATTR_ENTITY_ID
from homeassistant.core import (
    HomeAssistant,
    ServiceCall,
    ServiceResponse,
    SupportsResponse,
    callback,
)
from homeassistant.exceptions import ServiceValidationError
from homeassistant.helpers import config_validation as cv
import voluptuous as vol

from .const import ATTR_LED_CONFIG, ATTR_NAME, ATTR_NOTIFICATION, DOMAIN
from .orchestrator import LampieOrchestrator
from .types import (
    Color,
    Effect,
    InvalidColor,
    LEDConfig,
    LEDConfigSource,
    LEDConfigSourceType,
)

SERVICE_NAME_ACTIVATE: Final = "activate"
SERVICE_NAME_OVERRIDE: Final = "override"


def _led_config(
    value: Any,  # noqa: ANN401
) -> LEDConfig:
    try:
        return LEDConfig.from_config(value)
    except InvalidColor as e:
        msg = f"{value} is not a valid color; code: `{e.reason}`"
        raise vol.Invalid(msg) from e


def _led_configs(value: str | list | None) -> tuple[LEDConfig, ...] | None:
    if value is None:
        return value
    if isinstance(value, str):
        value = [value]
    return tuple(_led_config(item) for item in value)


SERVICE_SCHEMA_ACTIVATE: Final = vol.Schema(
    {
        vol.Required(ATTR_NOTIFICATION): vol.All(cv.ensure_list, [cv.string]),
        vol.Optional(ATTR_LED_CONFIG): _led_configs,
    },
)

SERVICE_SCHEMA_OVERRIDE: Final = vol.Schema(
    {
        vol.Required(ATTR_ENTITY_ID): cv.entity_ids,
        vol.Required(ATTR_LED_CONFIG): _led_configs,
        vol.Optional(ATTR_NAME): cv.string,
    },
)


async def _activate(
    call: ServiceCall,
    *,
    hass: HomeAssistant,
) -> ServiceResponse:
    slugs: list[str] = call.data[ATTR_NOTIFICATION]
    led_config: tuple[LEDConfig, ...] | None = call.data.get(ATTR_LED_CONFIG)
    led_config_source = (
        LEDConfigSource(f"{','.join(slugs)}[custom]", LEDConfigSourceType.NOTIFICATION)
        if led_config is not None
        else None
    )

    orchestrator: LampieOrchestrator = hass.data[DOMAIN]

    for slug in slugs:
        if not orchestrator.has_notification(slug):
            raise ServiceValidationError(
                translation_domain=DOMAIN,
                translation_key="invalid_notification",
                translation_placeholders={
                    "slug": slug,
                },
            )

    for slug in slugs:
        await orchestrator.activate_notification(  # waiting in loop to avoid flooding the network
            slug,
            led_config=led_config,
            led_config_source=led_config_source,
        )


async def _override(
    call: ServiceCall,
    *,
    hass: HomeAssistant,
) -> ServiceResponse:
    name: str = call.data.get(ATTR_NAME) or f"{DOMAIN}.{SERVICE_NAME_OVERRIDE}"
    switch_ids: list[str] = call.data[ATTR_ENTITY_ID]
    led_config: tuple[LEDConfig, ...] | None = call.data[ATTR_LED_CONFIG]
    led_config_source = LEDConfigSource(name, LEDConfigSourceType.OVERRIDE)

    if led_config is None:  # none indicates a reset
        led_config_source = LEDConfigSource(None)
    elif not led_config:  # convert empty led config to clear config
        led_config = (LEDConfig(Color.BLUE, Effect.CLEAR),)

    orchestrator: LampieOrchestrator = hass.data[DOMAIN]

    for switch_id in switch_ids:
        await orchestrator.override_switch(  # waiting in loop to avoid flooding the network
            switch_id,
            led_config_source=led_config_source,
            led_config=led_config,
        )


@callback
def async_setup_services(hass: HomeAssistant) -> None:
    """Set up Lampie services."""
    hass.services.async_register(
        DOMAIN,
        SERVICE_NAME_ACTIVATE,
        partial(_activate, hass=hass),
        schema=SERVICE_SCHEMA_ACTIVATE,
        supports_response=SupportsResponse.NONE,
    )

    hass.services.async_register(
        DOMAIN,
        SERVICE_NAME_OVERRIDE,
        partial(_override, hass=hass),
        schema=SERVICE_SCHEMA_OVERRIDE,
        supports_response=SupportsResponse.NONE,
    )
