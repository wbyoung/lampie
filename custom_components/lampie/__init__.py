"""The Lampie integration."""

from __future__ import annotations

import logging

from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers import config_validation as cv
from homeassistant.helpers.event import async_track_entity_registry_updated_event
from homeassistant.helpers.typing import ConfigType

from .const import CONF_SWITCH_ENTITIES, DOMAIN
from .coordinator import LampieUpdateCoordinator
from .orchestrator import LampieOrchestrator
from .services import async_setup_services
from .types import LampieConfigEntry, LampieConfigEntryRuntimeData

_LOGGER = logging.getLogger(__name__)

PLATFORMS: list[Platform] = [Platform.SENSOR, Platform.SWITCH]
CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)


async def async_setup(  # noqa: RUF029
    hass: HomeAssistant,
    config: ConfigType,  # noqa: ARG001
) -> bool:
    """Set up Lampie services.

    Returns:
        If the setup was successful.
    """
    async_setup_services(hass)

    return True


async def async_setup_entry(hass: HomeAssistant, entry: LampieConfigEntry) -> bool:
    """Set up Lampie from a config entry.

    Returns:
        If the setup was successful.
    """
    _LOGGER.debug("setup %s with config:%s", entry.title, entry.data)

    if DOMAIN not in hass.data:
        orchestrator = LampieOrchestrator(hass)
        hass.data[DOMAIN] = orchestrator

        await orchestrator.setup()

    coordinator = LampieUpdateCoordinator(hass, entry)
    orchestrator = hass.data[DOMAIN]
    await orchestrator.add_coordinator(coordinator)
    entry.runtime_data = LampieConfigEntryRuntimeData(
        orchestrator=orchestrator,
        coordinator=coordinator,
    )

    entry.async_on_unload(
        async_track_entity_registry_updated_event(
            hass,
            entry.data[CONF_SWITCH_ENTITIES],
            coordinator.async_handle_switch_id_change,
        ),
    )

    await coordinator.async_config_entry_first_refresh()
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
    entry.async_on_unload(entry.add_update_listener(_async_update_listener))
    return True


async def async_unload_entry(hass: HomeAssistant, entry: LampieConfigEntry) -> bool:
    """Unload a config entry.

    Returns:
        If the unload was successful.
    """
    coordinator: LampieUpdateCoordinator = entry.runtime_data.coordinator
    unload_ok: bool = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)

    if orchestrator := entry.runtime_data.orchestrator:
        await orchestrator.remove_coordinator(coordinator)

        if orchestrator.teardown() and orchestrator == hass.data.get(DOMAIN):
            hass.data.pop(DOMAIN)

    return unload_ok


async def _async_update_listener(
    hass: HomeAssistant,
    entry: LampieConfigEntry,
) -> None:
    """Handle options update."""
    runtime_data: LampieConfigEntryRuntimeData = entry.runtime_data

    if runtime_data.auto_reload_enabled:
        await hass.config_entries.async_reload(entry.entry_id)
