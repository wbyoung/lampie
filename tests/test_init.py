"""Test component setup."""

from unittest.mock import patch

from homeassistant.config_entries import ConfigEntryState
from homeassistant.core import HomeAssistant
from homeassistant.helpers import (
    device_registry as dr,
    entity_registry as er,
    issue_registry as ir,
)
from homeassistant.setup import async_setup_component
import pytest
from pytest_homeassistant_custom_component.common import MockConfigEntry
from syrupy.assertion import SnapshotAssertion
from syrupy.filters import props

from custom_components.lampie.const import (
    CONF_COLOR,
    CONF_EFFECT,
    CONF_PRIORITY,
    CONF_SWITCH_ENTITIES,
    DOMAIN,
)

from . import add_mock_switch, setup_integration


async def test_async_setup(hass: HomeAssistant):
    """Test the component gets setup."""
    assert await async_setup_component(hass, DOMAIN, {}) is True


async def test_standard_setup(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    switch: er.RegistryEntry,
    device_registry: dr.DeviceRegistry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test standard setup."""
    await setup_integration(hass, config_entry)

    device = device_registry.async_get(switch.device_id)

    assert device is not None
    assert device.id == switch.device_id
    assert device == snapshot(
        name="device",
        exclude=props(
            # compat for HA DeviceRegistryEntrySnapshot <2025.8.0 and >=2025.6.1
            "suggested_area",
            # compat for HA DeviceRegistryEntrySnapshot <2025.9.0 and >=2025.6.1
            "is_new",
        ),
    )


async def test_config_entries_linked_to_switch_device(
    hass: HomeAssistant,
    device_registry: dr.DeviceRegistry,
) -> None:
    """Test that the config entries are linked to the switch devices."""
    entryway_switch = add_mock_switch(hass, "light.entryway")
    kitchen_switch = add_mock_switch(hass, "light.kitchen")

    doors_open_entry = MockConfigEntry(
        domain=DOMAIN,
        title="Doors Open",
        entry_id="mock-doors-open-id",
        data={
            CONF_COLOR: "red",
            CONF_EFFECT: "open_close",
            CONF_SWITCH_ENTITIES: ["light.entryway"],
        },
    )
    doors_open_entry.add_to_hass(hass)
    await setup_integration(hass, doors_open_entry)
    doors_open_device = device_registry.async_get_device(
        {(DOMAIN, doors_open_entry.entry_id)}
    )

    medicine_entry = MockConfigEntry(
        domain=DOMAIN,
        title="Medicine",
        entry_id="mock-medicine-id",
        data={
            CONF_COLOR: "cyan",
            CONF_EFFECT: "slow_blink",
            CONF_SWITCH_ENTITIES: ["light.kitchen"],
        },
    )
    medicine_entry.add_to_hass(hass)
    await setup_integration(hass, medicine_entry)
    medicine_device = device_registry.async_get_device(
        {(DOMAIN, medicine_entry.entry_id)}
    )

    device_ids = {
        device.id
        for device in device_registry.devices.values()
        if device.config_entries & {"mock-doors-open-id", "mock-medicine-id"}
    }

    assert device_ids == {
        doors_open_device.id,
        medicine_device.id,
        entryway_switch.device_id,
        kitchen_switch.device_id,
    }


async def test_primary_config_entry_sensor_ownership(
    hass: HomeAssistant,
    device_registry: dr.DeviceRegistry,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test that sensor is not duplicated across entities on multiple entries."""
    entryway_switch = add_mock_switch(hass, "light.entryway")
    kitchen_switch = add_mock_switch(hass, "light.kitchen")

    doors_open_entry = MockConfigEntry(
        domain=DOMAIN,
        title="Doors Open",
        entry_id="mock-doors-open-id",
        data={
            CONF_COLOR: "red",
            CONF_EFFECT: "open_close",
            CONF_SWITCH_ENTITIES: ["light.entryway", "light.kitchen"],
            CONF_PRIORITY: {
                "light.kitchen": ["medicine", "doors_open"],
            },
        },
    )
    doors_open_entry.add_to_hass(hass)
    await setup_integration(hass, doors_open_entry)
    doors_open_device = device_registry.async_get_device(
        {(DOMAIN, doors_open_entry.entry_id)}
    )

    medicine_entry = MockConfigEntry(
        domain=DOMAIN,
        title="Medicine",
        entry_id="mock-medicine-id",
        data={
            CONF_COLOR: "cyan",
            CONF_EFFECT: "slow_blink",
            CONF_SWITCH_ENTITIES: ["light.kitchen"],
            CONF_PRIORITY: {
                "light.kitchen": ["medicine", "doors_open"],
            },
        },
    )
    medicine_entry.add_to_hass(hass)
    await setup_integration(hass, medicine_entry)
    medicine_device = device_registry.async_get_device(
        {(DOMAIN, medicine_entry.entry_id)}
    )

    doors_open_switch_entities = [
        entity
        for entity in entity_registry.entities.get_entries_for_config_entry_id(
            doors_open_entry.entry_id
        )
        if entity.device_id != doors_open_device.id
    ]
    for entity in doors_open_switch_entities:
        assert entity.device_id == entryway_switch.device_id

    medicine_switch_entities = [
        entity
        for entity in entity_registry.entities.get_entries_for_config_entry_id(
            medicine_entry.entry_id
        )
        if entity.device_id != medicine_device.id
    ]
    for entity in medicine_switch_entities:
        assert entity.device_id == kitchen_switch.device_id

    assert len(doors_open_switch_entities) == len(medicine_switch_entities)


async def test_mismatched_priorities_generate_warning(
    hass: HomeAssistant,
    switch: er.RegistryEntry,
    caplog: pytest.LogCaptureFixture,
) -> None:
    """Test that mismatched config priorities generate a warning."""
    doors_open_entry = MockConfigEntry(
        domain=DOMAIN,
        title="Doors Open",
        entry_id="mock-doors-open-id",
        data={
            CONF_COLOR: "red",
            CONF_EFFECT: "open_close",
            CONF_SWITCH_ENTITIES: ["light.kitchen"],
            CONF_PRIORITY: {
                "light.kitchen": ["medicine", "doors_open"],
            },
        },
    )
    doors_open_entry.add_to_hass(hass)
    await setup_integration(hass, doors_open_entry)

    medicine_entry = MockConfigEntry(
        domain=DOMAIN,
        title="Medicine",
        entry_id="mock-medicine-id",
        data={
            CONF_COLOR: "cyan",
            CONF_EFFECT: "slow_blink",
            CONF_SWITCH_ENTITIES: ["light.kitchen"],
            CONF_PRIORITY: {
                "light.kitchen": ["doors_open", "medicine"],
            },
        },
    )
    medicine_entry.add_to_hass(hass)
    await setup_integration(hass, medicine_entry)

    assert (
        "priorities mismatch found for light.kitchen in medicine: "
        "['medicine', 'doors_open'] (previously seen) != "
        "['doors_open', 'medicine'] (for medicine), loading order: 'doors_open'"
    ) in caplog.text


async def test_update_entry(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
) -> None:
    await setup_integration(hass, config_entry)

    with patch(
        "homeassistant.config_entries.ConfigEntries.async_reload",
        return_value=True,
    ) as mock_reload:
        hass.config_entries.async_update_entry(
            config_entry,
            data={"arbitrary-update": "value1"},
        )
        await hass.async_block_till_done()
        assert mock_reload.called

        mock_reload.reset_mock()
        config_entry.runtime_data.auto_reload_enabled = False
        hass.config_entries.async_update_entry(
            config_entry,
            data={"arbitrary-update": "value2"},
        )
        await hass.async_block_till_done()
        assert not mock_reload.called


async def test_unload(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
) -> None:
    await setup_integration(hass, config_entry)

    assert len(hass.config_entries.async_entries(DOMAIN)) == 1
    assert config_entry.state is ConfigEntryState.LOADED

    assert await hass.config_entries.async_unload(config_entry.entry_id)
    await hass.async_block_till_done()

    assert config_entry.state is ConfigEntryState.NOT_LOADED
    assert not hass.data.get(DOMAIN)


async def test_unload_failure(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
) -> None:
    await setup_integration(hass, config_entry)

    assert len(hass.config_entries.async_entries(DOMAIN)) == 1
    assert config_entry.state is ConfigEntryState.LOADED

    with patch(
        "homeassistant.config_entries.ConfigEntries.async_unload_platforms",
        return_value=False,
    ):
        await hass.config_entries.async_unload(config_entry.entry_id)
        await hass.async_block_till_done()

    assert config_entry.state is ConfigEntryState.FAILED_UNLOAD


async def test_renamed_switch_entity(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    switch: er.RegistryEntry,
    device_registry: dr.DeviceRegistry,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test that config entry data is updated when switch entity is renamed."""
    hass.states.async_set(switch.entity_id, "on")

    assert config_entry.data[CONF_SWITCH_ENTITIES] == [switch.entity_id]
    await setup_integration(hass, config_entry)

    entity_registry.async_update_entity(
        switch.entity_id,
        new_entity_id=f"{switch.entity_id}_renamed",
    )
    await hass.async_block_till_done()
    assert config_entry.data[CONF_SWITCH_ENTITIES] == [f"{switch.entity_id}_renamed"]


async def test_create_removed_switch_entity_issue(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    switch: er.RegistryEntry,
    issue_registry: ir.IssueRegistry,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test we create an issue for removed switch entities."""
    hass.states.async_set(switch.entity_id, "not_home")

    await setup_integration(hass, config_entry)

    hass.states.async_remove(switch.entity_id)
    entity_registry.async_remove(switch.entity_id)
    await hass.async_block_till_done()

    assert issue_registry.async_get_issue(
        DOMAIN,
        f"switch_entity_removed_{switch.entity_id}",
    )
