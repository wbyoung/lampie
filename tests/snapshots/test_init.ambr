# serializer version: 1
# name: test_standard_setup[device]
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'config_entries_subentries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'zha',
        'mock-ieee:kitchen',
      ),
    }),
    'labels': set({
    }),
    'manufacturer': 'Inovelli',
    'model': 'VZM31-SN',
    'model_id': None,
    'name': 'Kitchen',
    'name_by_user': None,
    'primary_config_entry': <ANY>,
    'serial_number': None,
    'sw_version': None,
    'via_device_id': None,
  })
# ---
